### PR TITLE
Prototype testing workflow skeleton

### PR DESCRIPTION
**Pull request by camerski**
### Description
Prototype of the test workflow. This is mostly stubbed-out code, but it shows the overall flow of reading a manifest file, setting up a test cluster, running `integtest.sh` for each component, and reporting results.

You can run the tool on a demo manifest that contains the OpenSearch 1.0.0 release artifact:

`python tester.py https://artifacts.opensearch.org/builds/opensearch/1.0.0/manifest.yml`

Make sure you have `pyyaml` installed (with `pip install pyyaml` or whatever package manager you prefer to use if you don't like pip).

The goal of this PR is to get feedback on the general approach, and to highlight the subtasks we should add to https://github.com/opensearch-project/opensearch-build/issues/122 for the next few days. Assuming people agree with this approach, I suggest the following subtasks in rough order of priority:

* Implement LocalTestCluster to stand up / tear down a test cluster on the local host.
* Implement IntegTestSuite so that it can capture test results and logs.
* Implement a mechanism for storing test results, logs, and metadata.
* Add failure handling and make sure `tester.py` cleans up (especially remote test clusters) and returns a suitable exit code.
* Add `tester.py` to a Jenkins or GitHub Actions workflow that can accept a manifest URL as an argument.
* Create and implement RemoteSingleNodeCluster to set up a remote single-node cluster using CDK.
* Create and implement RemoteMultiNodeCluster.
* Create and implement PerformanceTestSuite.
* Change `tester.py` to execute integ test suites in parallel.
 
### Issues Resolved
https://github.com/opensearch-project/opensearch-build/issues/122
 
### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
Meta question: do we have other workflows/tools that are Python. I know the signer is. Just wondering whether we are trying to standardize this in any way?

File: workflows/testing/test_workflow/integ_test.py
Comment (peterzhuamazon) [inline workflows/testing/test_workflow/integ_test.py:11]: -e?

**Comment by camerski** [comment]
> Meta question: do we have other workflows/tools that are Python. I know the signer is. Just wondering whether we are trying to standardize this in any way?

At this point in time we have not had a discussion about standardization. This PR should not be construed as an endorsement of Python :)

**Comment by setiah** [comment]
Thanks @camerski for proposing this. This is definitely one way to write orchestration layer but I am leaning towards a Jenkins pipeline based workflow due to a couple of reasons. It seems like in this approach, we might need to reinvent the wheel for a couple of great features that are available out of the box w/ Jenkins, for instance - 
1. Need to write our own state machine logic to run workflows.
2. Need to take dependency on DDB (or any other external service) to persist intermediate states. We will also need to add logic for DDB interactions. 
3. No visualization for workflows available
4. Need to write job parallelization logic inside this python workflow to run independent tests in parallel.
5. Not clear yet how we breakdown tests into separate jobs. If we choose to run everything as one job, it makes things too monolithic. This is a problem because when there is a job failure someone will need to actually dig into logs to check was it related to integ tests/ perf tests/ bwc or any other workflow that we run. A clear separation of jobs would help to better understand general characteristics such as which test suite is more flaky and needs fixing.

Jenkins pipeline seem to offer all of this and more out of the box. It seems to have easy way to pass params between stages.

File: workflows/testing/test_workflow/git.py
Comment (peternied) [inline workflows/testing/test_workflow/git.py:8]: Do we need to worry about non-public repositories needing alternative credentials?

File: workflows/testing/test_workflow/integ_test.py
Comment (peternied) [inline workflows/testing/test_workflow/integ_test.py:13]: This seems like an exceptional state, shouldn't all repositories include integration tests?  I'd recommend throwing with a nasty error so they have to be dealt with

File: workflows/testing/test_workflow/manifest.py
Comment (peternied) [inline workflows/testing/test_workflow/manifest.py:7]: hash? Not sure what this naming convention means?

File: workflows/testing/test_workflow/manifest.py
Comment (peternied) [inline workflows/testing/test_workflow/manifest.py:34]: Save some lines with `@dataclass` or something similar?

File: workflows/testing/tester.py
Comment (peternied) [inline workflows/testing/tester.py:27]: Why add this special case?

**Comment by camerski** [comment]
All good points, @setiah. Let me explain why I approached it this way.

> 1. Need to write our own state machine logic to run workflows.

Correct. Definitely a downside of this approach.

> 2. Need to take dependency on DDB (or any other external service) to persist intermediate states. We will also need to add logic for DDB interactions.

Correct, but not really a big deal. DDB is pretty simple.

> 3. No visualization for workflows available

Not sure how much this matters.

> 4. Need to write job parallelization logic inside this python workflow to run independent tests in parallel.

This is not difficult.

> 5. Not clear yet how we breakdown tests into separate jobs. If we choose to run everything as one job, it makes things too monolithic. This is a problem because when there is a job failure someone will need to actually dig into logs to check was it related to integ tests/ perf tests/ bwc or any other workflow that we run. A clear separation of jobs would help to better understand general characteristics such as which test suite is more flaky and needs fixing.

This is where the fun starts :)

Jenkins pipelines are static - there is not (or, at least, I couldn't find) a way to dynamically change the number of stages at runtime. This is important for test runs: If there are `n` components in the bundle then we want to run `n` test suites, if the next bundle has `n + 1` components then the pipeline needs to be able to add a parallel step for the `n + 1th` component, without needing a code change for the pipeline.

That basically means that we need to build our own parallelism - the best the pipeline can do is have a step that executes `run-all-integ-tests.sh`. The other steps in the workflow (download and parse the manifest, check out all the repos) are pretty trivial and I'm not concerned about being able to restart a failed workflow for those pieces.

Test cluster lifecycle is also a bit tricky. I think we could manage that state using the Jenkins pipeline data storage model but I haven't prototyped that so it's unclear to me how well that will work, and what the sharp edges will be. I suspect we will need to store things like CloudFormation stack ARNs in external storage so that we have an easy way to terminate them if the Jenkins pipeline explodes.

All that said, I'm not opposed to doing it all in Jenkins. Let's prototype both and see where we land.

File: workflows/testing/tester.py
Comment (camerski) [reply workflows/testing/tester.py:27]: Because we don't have integtest.sh for this repo. It's a big repo, so checking out a ton of code that we won't use is a substantial waste of time.

File: workflows/testing/test_workflow/manifest.py
Comment (camerski) [reply workflows/testing/test_workflow/manifest.py:34]: My python-fu is weak :) Thanks!

File: workflows/testing/test_workflow/manifest.py
Comment (camerski) [reply workflows/testing/test_workflow/manifest.py:7]: It means the author of this code comes from a Ruby background. What's the convention in Python?

File: workflows/testing/test_workflow/integ_test.py
Comment (camerski) [reply workflows/testing/test_workflow/integ_test.py:13]: Eventually they should all include integration tests. We're not there yet, however.

File: workflows/testing/test_workflow/git.py
Comment (camerski) [reply workflows/testing/test_workflow/git.py:8]: I'm happy to say we don't support non-public repositories.

File: workflows/testing/test_workflow/integ_test.py
Comment (peternied) [reply workflows/testing/test_workflow/integ_test.py:13]: I tend to lend on not included in components into bundling until they meet a basic compliance level.  Forces earlier investment via components and makes our 'management' of onboarding new components simpler.

Otherwise we have to scan the logs for this error message that means we have an untested component.

/2cents

File: workflows/testing/test_workflow/git.py
Comment (dblock) [inline workflows/testing/test_workflow/git.py:15]: This doesn't fail when the command fails.

```python
import os
import tempfile
import subprocess

class GitRepository:
    def __init__(self, url, ref):
        self._url = url
        self._dir = tempfile.TemporaryDirectory()
        self.execute(f'git clone --branch {ref} --depth 1 {url} {self._dir.name}')
        print(f'Checked out {url}@{ref} into {self._dir.name}')

    def dir(self):
        return self._dir.name

    def execute(self, command):
        current = os.getcwd()
        print(f'Executing "{command}" in {self._dir.name}')
        os.chdir(self._dir.name)
        subprocess.check_call(command, shell = True)
        os.chdir(current)
```

File: workflows/testing/test_workflow/integ_test.py
Comment (dblock) [reply workflows/testing/test_workflow/integ_test.py:11]: This doesn't fail when integ tests fail.

File: workflows/testing/test_workflow/manifest.py
Comment (peternied) [reply workflows/testing/test_workflow/manifest.py:7]: ohhh... I have no idea - hash to me means md5 or sha1...  Blob or data come to mind, maybe whatever yaml is called, if it was json i'd call it json.

File: workflows/testing/tester.py
Comment (peternied) [reply workflows/testing/tester.py:27]: > checking out a ton of code that we won't use

Hmm... is this worth doing a sparse checkout to make it faster? What kind of performance should we expect?

**Comment by dblock** [comment]
I think for git checkout you can save yourself a lot of time and only check out depth 1.

```python
import os
import tempfile
import subprocess

class GitRepository:
    def __init__(self, url, ref):
        self._url = url
        self._dir = tempfile.TemporaryDirectory()
        self.execute(f'git init')
        self.execute(f'git remote add origin {url}')
        self.execute(f'git fetch --depth 1 origin {ref}')
        self.execute(f'git checkout FETCH_HEAD')
        print(f'Checked out {url}@{ref} into {self._dir.name}')

    def dir(self):
        return self._dir.name

    def execute(self, command):
        current = os.getcwd()
        print(f'Executing "{command}" in {self._dir.name}')
        os.chdir(self._dir.name)
        try:
            subprocess.check_call(command, shell = True)
        finally:
            os.chdir(current)
```

File: workflows/testing/test_workflow/integ_test.py
Comment (camerski) [reply workflows/testing/test_workflow/integ_test.py:11]: Yup, it's a prototype. Fault tolerance will come if/when we turn this into a production system.

File: workflows/testing/test_workflow/integ_test.py
Comment (camerski) [reply workflows/testing/test_workflow/integ_test.py:13]: Directionally, I agree. Since this is a prototype I'm happy to worry about that later.

**Comment by camerski** [comment]
> I think for git checkout you can save yourself a lot of time and only check out depth 1. Here's a full class with error handling.

Oh, yeah, depth 1 is *way* faster...thanks!

File: workflows/testing/test_workflow/git.py
Comment (camerski) [reply workflows/testing/test_workflow/git.py:15]: I didn't know about the `subprocess` library...that looks much better than `os.system`. Again, my python-fu is weak, so thanks for the pointer :)

File: workflows/testing/test_workflow/manifest.py
Comment (dblock) [reply workflows/testing/test_workflow/manifest.py:7]: Should just be `data`, e.g. https://github.com/opensearch-project/opensearch-build/blob/c9885c0f7fe4fb6808cf3b34364ef91efcdeb825/tools/bundle-build/lib/manifest.py

File: workflows/testing/test_workflow/git.py
Comment (dblock) [reply workflows/testing/test_workflow/git.py:15]: Beyond Python, make sure every method called fails loudly.

Comment (camerski) [reply workflows/testing/test_workflow/git.py:15]: Yes, of course. Remember that "the goal of this PR is to get feedback on the general approach, and to highlight the subtasks we should add to #122 for the next few days". Fault tolerance is not so important for prototype code.

File: workflows/testing/tester.py
Comment (camerski) [reply workflows/testing/tester.py:27]: Changed to use `depth 1`.

File: workflows/testing/test_workflow/manifest.py
Comment (camerski) [reply workflows/testing/test_workflow/manifest.py]: Done

File: workflows/testing/test_workflow/manifest.py
Comment (dblock) [reply workflows/testing/test_workflow/manifest.py:32]: Here's a similar refactor: https://github.com/opensearch-project/opensearch-build/pull/150

**Comment by dblock** [comment]
Let's also see if we can reuse the common parts, such as git and manifest implementations?

**Comment by camerski** [comment]
> Let's also see if we can reuse the common parts, such as git and manifest implementations?

Yeah, we should do a big refactoring exercise once we have the overall workflow more-or-less ready.

Once this PR is approved I can get started on some of the refactoring and extracting of common functionality.

**Comment by dblock** [comment]
> > Let's also see if we can reuse the common parts, such as git and manifest implementations?
> 
> Yeah, we should do a big refactoring exercise once we have the overall workflow more-or-less ready.
> 
> Once this PR is approved I can get started on some of the refactoring and extracting of common functionality.

You don't have to wait. Those .py classes that were written as part of #150 and #145 are directly reusable. Save yourself time rewriting/copying code.

File: workflows/testing/test_workflow/manifest.py
Comment (setiah) [inline workflows/testing/test_workflow/manifest.py:3]: We can separate this out as a separate tool to read/write manifest file

**Comment by camerski** [comment]
> > > Let's also see if we can reuse the common parts, such as git and manifest implementations?
> > 
> > 
> > Yeah, we should do a big refactoring exercise once we have the overall workflow more-or-less ready.
> > Once this PR is approved I can get started on some of the refactoring and extracting of common functionality.
> 
> You don't have to wait. Those .py classes that were written as part of #150 and #145 are directly reusable. Save yourself time rewriting/copying code.

Those changes in #150 and #145 also need to be refactored (calling a library `lib` is not a great idea). I would rather get this already-written code checked in so we can move fast on implementing the TestCluster and TestSuite classes, and refactor later.

File: workflows/testing/test_workflow/manifest.py
Comment (camerski) [reply workflows/testing/test_workflow/manifest.py:3]: Yup, there will be a big refactoring exercise later to unify this code with the build workflow, and pull out common components. Code re-use is not top priority right now; working prototype is the goal.

**Review by peternied — CHANGES_REQUESTED** [review]
I am happy to leave many issues open associated with this PR if we move it into a `prototype` folder so that we are sure what support level is associated with this code

**Comment by camerski** [comment]
> I am happy to leave many issues open associated with this PR if we move it into a `prototype` folder so that we are sure what support level is associated with this code

Sure, I can do that.

**Comment by setiah** [comment]
> Jenkins pipelines are static - there is not (or, at least, I couldn't find) a way to dynamically change the number of stages at runtime. This is important for test runs: If there are `n` components in the bundle then we want to run `n` test suites, if the next bundle has `n + 1` components then the pipeline needs to be able to add a parallel step for the `n + 1th` component, without needing a code change for the pipeline.
> 
> That basically means that we need to build our own parallelism - the best the pipeline can do is have a step that executes `run-all-integ-tests.sh`. The other steps in the workflow (download and parse the manifest, check out all the repos) are pretty trivial and I'm not concerned about being able to restart a failed workflow for those pieces.
 
Actually we can do parallelization at multiple levels. The first level is orchestration layer where Jenkins jobs can be made to run in parallel via pipeline. Orchestration framework can kick off additional parallel jobs for `n+1th` component. 
Another layer could be python tool itself, where the Jenkins orchestration can be made agnostic and the tool can have logic to parallelize multiple tests based on the bundle components. The possibility exists. We can come up with exact structure later when we are scaling test infra to support multiple bundles.

**Comment by dblock** [comment]
> > I am happy to leave many issues open associated with this PR if we move it into a `prototype` folder so that we are sure what support level is associated with this code
> 
> Sure, I can do that.

I think this is unnecessarily moving stuff around. We are building a system that needs to work, AFAIK the criteria to merge this code is that it completes an end-to-end test successfully, code improvements and refactorings are should have's.

**Comment by dblock** [comment]
> > Jenkins pipelines are static - there is not (or, at least, I couldn't find) a way to dynamically change the number of stages at runtime. This is important for test runs: If there are `n` components in the bundle then we want to run `n` test suites, if the next bundle has `n + 1` components then the pipeline needs to be able to add a parallel step for the `n + 1th` component, without needing a code change for the pipeline.
> > That basically means that we need to build our own parallelism - the best the pipeline can do is have a step that executes `run-all-integ-tests.sh`. The other steps in the workflow (download and parse the manifest, check out all the repos) are pretty trivial and I'm not concerned about being able to restart a failed workflow for those pieces.
> 
> Actually we can do parallelization at multiple levels. The first level is orchestration layer where Jenkins jobs can be made to run in parallel via pipeline. Orchestration framework can kick off additional parallel jobs for `n+1th` component.
> Another layer could be python tool itself, where the Jenkins orchestration can be made agnostic and the tool can have logic to parallelize multiple tests based on the bundle components. The possibility exists. We can come up with exact structure later when we are scaling test infra to support multiple bundles.

Is it a crazy idea to generate a Jenkins pipeline from a manifest?

**Comment by camerski** [comment]
> code improvements and refactorings are should have's and can be deferred, but not forgotten. That said, we ought to keep ourselves to a high bar for code quality during development, and not defer what can be done right away.

@dblock Given that statement, what exactly in this PR do you consider to be a must-have before approving, and what do you think can be deferred?

**Review by mch2 — APPROVED** [review]

**Review by peternied — APPROVED** [review]

**Comment by dblock** [comment]
> > code improvements and refactorings are should have's and can be deferred, but not forgotten. That said, we ought to keep ourselves to a high bar for code quality during development, and not defer what can be done right away.
> 
> @dblock Given that statement, what exactly in this PR do you consider to be a must-have before approving, and what do you think can be deferred?

You got two approvals so go for it.

**Pull request merged**
